distributed/tests/test_steal.py

```diff
@@ -110,11 +111,122 @@ async def test_worksteal_many_thieves(c, s, *workers):
     assert sum(map(len, s.has_what.values())) < 150
 
 
-@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)] * 2)
-async def test_dont_steal_unknown_functions(c, s, a, b):
-    futures = c.map(inc, range(100), workers=a.address, allow_other_workers=True)
-    await wait(futures)
-    assert len(a.data) >= 95, [len(a.data), len(b.data)]
+@gen_cluster(
+    client=True,
+    nthreads=[("127.0.0.1", 1)] * 2,
+    config={"distributed.scheduler.work-stealing-interval": "10ms"},
+)
+async def test_stop_plugin(c, s, a, b):
+    steal = s.extensions["stealing"]
+
+    await steal.stop()
+    futs = c.map(slowinc, range(10), workers=[a.address], allow_other_workers=True)
+    await c.gather(futs)
+    assert len(a.data) == 10
+
+    # nothing happens
+    for _ in range(10):
+        await steal.stop()
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("127.0.0.1", 1)] * 2,
+    config={"distributed.scheduler.work-stealing-interval": "1ms"},
+)
+async def test_stop_in_flight(c, s, a, b):
+    steal = s.extensions["stealing"]
+    num_tasks = 10
+    futs = c.map(
+        slowinc, range(num_tasks), workers=[a.address], allow_other_workers=True
+    )
+    while not steal.in_flight:
+        await asyncio.sleep(0)
+
+    assert steal.in_flight
+    await steal.stop()
+    assert not steal.in_flight
+
+    assert len(a.data) != num_tasks
+    del futs
+    while s.tasks or a.tasks or b.tasks:
+        await asyncio.sleep(0.1)
+
+    futs = c.map(
+        slowinc, range(num_tasks), workers=[a.address], allow_other_workers=True
+    )
+    await c.gather(futs)
+    assert len(a.data) == num_tasks
+
+    del futs
+    while s.tasks or a.tasks or b.tasks:
+        await asyncio.sleep(0.1)
+    futs = c.map(
+        slowinc, range(num_tasks), workers=[a.address], allow_other_workers=True
+    )
+    while not len(a.tasks) == num_tasks:
+        await asyncio.sleep(0.01)
+    assert len(b.tasks) == 0
+    await steal.start()
+    await c.gather(futs)
+    assert len(a.tasks) != num_tasks
+    assert len(b.tasks) != 0
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("127.0.0.1", 1)] * 2,
+    config={"distributed.scheduler.work-stealing-interval": "10ms"},
+)
+async def test_allow_tasks_stolen_before_first_completes(c, s, a, b):
+    # https://github.com/dask/distributed/issues/5564
+    from distributed import Semaphore
+
+    steal = s.extensions["stealing"]
+    await steal.stop()
+    lock = await Semaphore(max_leases=1)
```

Comment:
Why a Semaphore instead of a Lock?

Reply:
I had issues with the lock and have had some in the past. It somehow didn't lock. I may have forgotten to await _something_, idk. fwiw, I think we should throw lock away and replace it with the semaphore implementation.

Reply:
`Lock = lambda: Semaphore(max_leases=1)` seems reasonable to me. Nice to have fewer codepaths.

Reply:
I would do something like the following but that's a matter of taste :)

```
class Lock(Semaphore):
    def __init__(*args, **kwargs):
        super().__init__(*args, **kwargs, max_leases=1)
```
The semaphore also implements lease timeouts such that it handles dying workers gracefully by releasing the lock.

IIRC, the reason why I didn't do this from the beginning is that there are subtly differences in API causing the Lock API to break. For instance, you need to await a semaphore object before using it while you don't need to do this on a lock, etc. Also a lot of the tests are specific about the extension. However, we might probable just be able to delete all Lock specific tests since the semaphore should cover it anyhow.